### PR TITLE
Improve monitor listener main loop.

### PR DESCRIPTION
**Pull request by DimCitus**
The new code consumes less CPU and should react quicker to notifications
being available on the monitor's connection. Also, we integrate checks for
Postgres to be running in the main listener loop, retrying to connect to the
database when it's not ready yet, rather than bailing out to the main
supervision process.

See #386.

### CODE REVIEW

File: src/bin/pg_autoctl/monitor.c [hunk 1]
```diff
@@ -2994,16 +2994,54 @@ monitor_stop_maintenance(Monitor *monitor, int nodeId)
...
+	if (!block_signals(&mask, &orig_mask))
+	{
+		return false;
+	}
+
+	if (asked_to_stop || asked_to_stop_fast || asked_to_reload || asked_to_quit)
+	{
+		/* restore signal masks (un block them) now */
+		(void) unblock_signals(&orig_mask);
+		return false;
```
Comment (JelteF) [inline src/bin/pg_autoctl/monitor.c:3027]: I think this should be:
```suggestion
		return true;
```

**Review by JelteF — APPROVED** [review]
Some small comments, but other than this it seems fine.

File: src/bin/pg_autoctl/monitor.c [hunk 2]
```diff
@@ -2994,16 +2994,54 @@ monitor_stop_maintenance(Monitor *monitor, int nodeId)
...
+
 	fd_set input_mask;
 
-	if (connection == NULL)
+	/* block signals now: process them as if received during the pselect call */
+	if (!block_signals(&mask, &orig_mask))
+	{
+		return false;
+	}
+
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c:3022]: Thanks, applied now.

File: src/bin/pg_autoctl/monitor.c [hunk 3]
```diff
@@ -3013,19 +3051,28 @@ monitor_get_notifications(Monitor *monitor)
...
+		(void) unblock_signals(&orig_mask);
+
 		return false;   /* shouldn't happen */
 	}
+
 	FD_ZERO(&input_mask);
 	FD_SET(sock, &input_mask);
 
-	if (select(sock + 1, &input_mask, NULL, NULL, NULL) < 0)
+	ret = pselect(sock + 1, &input_mask, NULL, NULL, &timeout, &orig_mask);
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c]: Renamed those variables to `sig_mask` and `sig_mask_orig`.

File: src/bin/pg_autoctl/service_monitor.c [hunk 4]
```diff
@@ -274,18 +259,44 @@ monitor_service_run(Monitor *monitor)
...
+				/* maybe we failed to connect to the monitor */
+				if (monitor->pgsql.status != PG_CONNECTION_OK)
+				{
+					/* leave some time to the monitor before we try again */
+					sleep(PG_AUTOCTL_MONITOR_SLEEP_TIME);
+
+					if (firstLoop)
+					{
+						firstLoop = false;
+					}
```
[the remark this answers is not in the document]
Comment (DimCitus) [reply src/bin/pg_autoctl/service_monitor.c:271]: That way it looks like in the other main loop for the node-active client, and also allows to later maybe do more things. Also we needed to do that before the call to `monitor_get_notifications` where we `continue` when this returns false, so it's now moved with another action needed when firstLoop. Looks better accidentally now.

File: src/bin/pg_autoctl/monitor.c (hunk 1 above)
Comment (DimCitus) [reply src/bin/pg_autoctl/monitor.c:3031]: When we have received a termination signal, it's could be that we want to say that the function did run successfully, but in the case of a reload we are going to disconnect, do things, and then reconnect. So we're going to miss notifications. I think it's fair to return false when we have signals to process in the main loop.

``` c
		if (!monitor_get_notifications(monitor,

		                               /* we want the time in milliseconds */
									   PG_AUTOCTL_MONITOR_SLEEP_TIME * 1000))
		{
			log_warn("Re-establishing connection. We might miss notifications.");
			pgsql_finish(&(monitor->pgsql));

			continue;
		}
```

**Pull request merged**